File: array/387-First-Unique-Character-in-a-String.js
/*
387.First Unique Character in a String
https://leetcode.com/problems/first-unique-character-in-a-string/
12 July 2023
David Pratt
*/

/**
 * @param {string} s
 * @return {number}
 */

var firstUniqChar = function(s) {
    const sArr = s.split('');
    
    let hashMap = {};

    for(let i = 0; i < s.length; i += 1){
        let key = sArr[i];
        if(!hashMap[key]) hashMap[key] = 1;
        else hashMap[key] += 1;
    }

    for(let i = 0; i < sArr.length; i += 1){
        let key = sArr[i]
        if(hashMap[key] === 1) return i; 
    }

    return -1;
};

File: hash-map/387-First-Unique-Character-In-a-String.py
/*
387. First Unique Character in a String
https://leetcode.com/problems/first-unique-character-in-a-string/
3 December 2023
David Pratt
*/

class Solution:
    def firstUniqChar(self, s: str) -> int:
        dict = {}

        for letter in s:
            if letter in dict:
                dict[letter] +=1
            else:
                dict[letter] = 1

        for i in range(len(s)):
            if(s[i] != s[i + 1] & dict[s] == 1):
                return i
        return -1
        
